### PR TITLE
Add glean test API documentation [ci skip] 

### PR DESCRIPTION
**Pull request by travis79**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/service/glean/README.md [hunk 1]
```diff
@@ -129,6 +129,75 @@ adb shell am start -n org.mozilla.samples.glean/mozilla.components.service.glean
   --es sendPing metrics
 ```
 
+## Unit testing glean metrics
```
Comment (georgf) [inline components/service/glean/README.md]: How about...
- we instead add a short section on testing below [adding new metrics](https://github.com/mozilla-mobile/android-components/blob/master/components/service/glean/README.md#adding-new-metrics)
- we move this detailed information to e.g. `docs/metrics/testing-metrics.md`

Comment (georgf) [reply components/service/glean/README.md]: As is this is rather long for an overview README.

Comment (travis79) [reply components/service/glean/README.md]: Good idea!  I just pushed this change, adding a short summary in the README.md with a link to `docs/metrics/testing-metrics.md` which now contains the full description and examples.

File: components/service/glean/docs/metrics/testing-metrics.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+In order to support unit testing inside of client applications using glean, a set of testing API 
+functions have been included.  The intent is to make glean easier to test 'out of the box' in any
+client application it may be used in.  These functions expose a way to inspect and validate recorded
+metric values within the client application but are restricted to test code only through visibility
+annotations (`@VisibleForTesting(otherwise = VisibleForTesting.NONE)`).
+
+## General test API method semantics
+
+To check if a value exists (i.e. it has been recorded), there is a `testHasValue(pingName: String)`
```
Comment (georgf) [inline components/service/glean/docs/metrics/testing-metrics.md]: Can we note here already that `pingName` is optional?
Or simply use `testHasValue()` and don't mention `pingName` until later in the document?

Most users should not need to use `pingName` and might get distracted by this.

File: components/service/glean/docs/metrics/testing-metrics.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+To check the actual values, there is a `testGetValue(pingName: String)` function on each of the
+metric instances.  This function will return a datatype appropriate to the specific type of the
+metric it is being used with.
+
+Example:
+```kotlin
+GleanMetrics.Foo.UriCount.testGetValue()
+```
+
+These functions require a `pingName` arugument.  This is the name of the ping that the metric is
```
Comment (georgf) [inline components/service/glean/docs/metrics/testing-metrics.md]: Following up to above, can we make that a separate sub section, *Testing metrics for custom pings*?

File: components/service/glean/docs/metrics/testing-metrics.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+being sent in, such as `baseline`, `events`, or `metrics`.  This will automatically default to the
+"default" ping that the metric is sent in.  A user should only need to provide a `pingName` if the
```
Comment (georgf) [inline components/service/glean/docs/metrics/testing-metrics.md]: Can you provide an example for *default*? Say *"(e.g. the `"events"` ping for events")*.

File: components/service/glean/docs/metrics/testing-metrics.md (hunk 3 above)
Comment (georgf) [reply components/service/glean/docs/metrics/testing-metrics.md]: Below the `testClearAllData()` section probably though?

File: components/service/glean/docs/metrics/testing-metrics.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+GleanMetrics.Foo.UriCount.testHasValue()
+```
+
+To check the actual values, there is a `testGetValue()` function on each of the metric instances.  
+This function will return a datatype appropriate to the specific type of the metric it is being used 
+with.
+
+Example:
+```kotlin
+GleanMetrics.Foo.UriCount.testGetValue()
```
Comment (georgf) [inline components/service/glean/docs/metrics/testing-metrics.md]: Nit: Add something that makes clear that a value is returned, like `... == 5`?

File: components/service/glean/docs/metrics/testing-metrics.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+them inside of client code.
+
+## Testing metrics for custom pings
+
+In order to test metrics where the metric is included in more than one ping, the test functions take 
+an optional `pingName` argument.  This is the name of the ping that the metric is being sent in, 
+such as `events` for the "events" ping, or `metrics` for the "metrics" ping.  This could also be a 
+custom ping name that the metric is being sent in.  In most cases you should not have to supply the 
+ping name to the test function and can just use the default which is the "default" ping that this 
+metric is sent in.  A user should only need to provide a `pingName` if the metric is being sent in
```
Comment (georgf) [inline components/service/glean/docs/metrics/testing-metrics.md]: s/A user/You/

File: components/service/glean/docs/metrics/testing-metrics.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+To check the actual values, there is a `testGetValue()` function on each of the metric instances.  
+This function will return a datatype appropriate to the specific type of the metric it is being used 
+with.
+
+Example:
+```kotlin
+GleanMetrics.Foo.UriCount.testGetValue()
+```
+
```
Comment (georgf) [inline components/service/glean/docs/metrics/testing-metrics.md]: I think this is missing `enableTestingMode()`?

File: components/service/glean/docs/metrics/testing-metrics.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+## General test API method semantics
+
+In order to prevent issues with async calls when unit testing glean, it is important to put glean
+into testing mode by calling `Glean.enableTestingMode()`.  This will ensure that metrics are done
+recording when the other test functions are used.
+
+To check if a value exists (i.e. it has been recorded), there is a `testHasValue()` function on each 
+of the metric instances.
+
+Example:
```
Comment (georgf) [inline components/service/glean/docs/metrics/testing-metrics.md]: Last nit, i promise!
How about removing all the "Example:" lines? You can save some whitespace etc. by using:
> To check if a value exists (i.e. it has been recorded), there is a `testHasValue()` function on each 
of the metric instances:
> ```kotlin
> assertTrue(GleanMetrics.Foo.UriCount.testHasValue())
> ```

**Pull request merged**